kedro/framework/project/__init__.py

```diff
@@ -259,3 +261,58 @@ def validate_settings():
         )
 
     importlib.import_module(f"{PACKAGE_NAME}.settings")
+
+
+def find_pipelines() -> Dict[str, Pipeline]:
+    """Automatically find modular pipelines having a ``create_pipeline``
+    function. By default, projects created using Kedro 0.18.3 and higher
+    call this function to autoregister pipelines upon creation/addition.
+
+    Returns:
+        A generated mapping from pipeline names to ``Pipeline`` objects.
+
+    Warns:
+        UserWarning: When a module does not expose a ``create_pipeline``
+            function, the ``create_pipeline`` function does not return a
+            ``Pipeline`` object, or if the module import fails up front.
+    """
+    pipelines_dict = {"__default__": pipeline([])}
+    for pipeline_dir in importlib_resources.files(
+        f"{PACKAGE_NAME}.pipelines"
+    ).iterdir():
```

Comment:
I don't know `importlib_resource` well enough, but this seems actually iterating the filesystem.
```
importlib_resources.files(
        f"{PACKAGE_NAME}.pipelines"
    )
```

If this is the case we need to break it down to do error-handling when `importlib_resources.files(
        f"{PACKAGE_NAME}.pipelines"
    )` return `ModuleNotFoundError`

Reply:
@noklam Sorry, I don't think I follow what you're suggesting here...

Reply:
@deepyaman Sorry for the late response, GH notifications don't work very well sometimes.

I think this needs to be fixed before this PR is merged. The code block below is outside of the `try/catch` block and it will stop the program running if the `pipelines` folder isn't exist due to `ModuleNotFoundError`.

```python
pipeline_dir in importlib_resources.files(
        f"{PACKAGE_NAME}.pipelines"
    ).iterdir():
```

This assume `importlib_resources.files( f"{PACKAGE_NAME}.pipelines" )` always return a vaild `Path` object, thus the subsequent `iterdir()` method call. However, if the `f"{PACKAGE_NAME}.pipelines"` doesn't exist, you will get a `ModuleNotFoundError` instead. `ModuleNotFoundError.iterdir()` will be a bug and not handled correctly with the current logic.

Some of our starters only have a `package_name/pipeline.py` file but not a `package_name/pipelines/pipeline.py`

To fix this, we probably need something like this
```python
try: 
  module = importlib_resources.files(f"{PACKAGE_NAME}.pipelines")

except ModuleNotFoundError:  # Or maybe some more?
  logger.info(f"{PACKAGE_NAME}.pipelines does not exists")

else:
   # Continue with the iterations...
   for pipeline_dir in module.iterdir():
       ...
```

Reply:
Based on https://github.com/kedro-org/kedro/pull/1706#issuecomment-1207937897, I assumed that it would be OK to not support for now:

> However, since this single file (compared to multiple directories) is a bit of an unusual/special case I don't mind too much if we just don't support it. Or we can come back and add support for it later. it does seem like a good case to support though since it's the "simplified kedro project structure" that I think we should still support.

I don't think we should swallow the `ModuleNotFoundError`, since `find_pipelines()` (currently) does expect a `pipelines` directory it can iterate over. That being said, it's also a choose-to-use feature, that's enabled by default when you do `kedro new`. Something with a simplified structure without a `pipelines` directory would simply not use this.

I personally am not even convinced that the simplified structure needs to be supported. If you just have a `pipeline.py` file, with the one `create_pipeline()` method in there, why do you need a magical `find_pipelines()`? The value of find pipelines is that you don't need to register each pipeline when you create it; in this case, just be explicit and define the one pipeline in the registry.

Reply:
@deepyaman Thanks for the comment! I think you are right that the error shouldn't be swallowed.

My view is slightly different:
1. I have a stronger view that we should support the simple `pipeline.py` case.
2. The error should still be handled and throw a more explicit error, since the module is not imported by the user, it's not obvious why it is failing. It's also better if we mentioned this expected structure in the docstring 

I mainly see this feature as a **beginner-friendly** feature, something that "just work" even if you don't understand how pipeline registration works, as long as you are creating a project with `kedro new` and `kedro create pipelines` (no need to edit the `pipeline_registry.py`).

We also have [pandas-iris](https://github.com/kedro-org/kedro-starters/tree/main/pyspark-iris/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D) and other `iris` starters that actually follow this simplified structure. It is strange to me a beginner-friendly feature is not available for a simple project.

```python
def register_pipelines() -> Dict[str, Pipeline]:
    """Register the project's pipelines.
    Returns:
        A mapping from a pipeline name to a ``Pipeline`` object.
    """

    return {
        "__default__": create_pipeline(),
    }
```
If we support case 1, then 99% of the cases are covered and we won't see the error very often (Point 2) unless users are doing some custom thing (not using standard starters or they have custom structure/logic), which is okay to throw an error. 

Imagine a beginner that starts with the `pandas-iris` project, and he found `modular_pipeline` later, how likely is the user going to find out the `find_pipelines` function afterward? It would be a smoother experience if the pipeline is discoverable after he does `kedro pipeline create`. It's probably because I always view `starters` and Kedro's CLI as entry points for feature discovery.

I think a unified workflow is quite important, especially if we are focusing on expanding the user base with less software engineering background. We shouldn't try too hard to explain.

Reply:
@noklam What's the user workflow if they start with the simplified structure, then `kedro pipeline create another_one`? Would the expected behavior be to include pipelines from both `pipeline.py` and `pipelines/another_one/pipeline.py`? Or would they refactor the `pipeline.py` into `pipelines/my_first/pipeline.py`?

Reply:
@deepyaman Ideally they should be refactored into folders at the end, but I don't think it's unreasonable to combine both of them as `__default__`.

In that case, `find_pipelines` is always there and users don't have to update this block, and they can also do `kedro run --pipeline=modular_pipeline`.

```python
  pipelines = find_pipelines()
  pipelines["__default__"] = sum(pipelines.values())
  return pipelines
```

[Antony has a different view that](https://github.com/kedro-org/kedro/pull/1706#issuecomment-1207937897) it should stop as soon as `pipeline.py` is found.

Reply:
Very good discussion here! I started 👍 the ones that I agreed with and that ended up being all the points made. I'll make a new issue so we can discuss this further. 

For the time being I'm happy to merge this without supporting the simple project structure (single `pipeline.py` file) case at all.

@deepyaman please could you update the starters which do have the multiple pipeline structure (probably just `spaceflights` from memory) to use the new `find_pipelines` functionality? 🙂

Reply:
See https://github.com/kedro-org/kedro/issues/1812.

Reply:
@AntonyMilneQB Thank you for chiming in. I'm in the process of updating to support single pipelines, but would also be more than happy to merge first and push it as a separate PR.

> @deepyaman please could you update the starters which do have the multiple pipeline structure (probably just `spaceflights` from memory) to use the new `find_pipelines` functionality? 🙂

Doesn't this need to be released first?

Reply:
Yes indeed, this must be released first - just wanted to mention it while I remembered.